plugins/exaile-shuffle-collection/__init__.py

```diff
@@ -0,0 +1,135 @@
+#!/usr/bin/python
+
+# Copyright (C) 2012 Rainer Hihn ( rainer@hihn.org )
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 1, or (at your option)
+# any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
+
+import logging
+import xl
+import xlgui
+import xl.player.adapters
+
+from xlgui import main
+
+logger = logging.getLogger(__name__)
+SHUFFLE = None
+
+def enable(exaile):
+    """Enables the Shuffle Collection plugin
+    """
+    global SHUFFLE
+    SHUFFLE = Shuffle(exaile)
+    if exaile.loading:
+        xl.event.add_callback(_enable, 'gui_loaded')
+    else:
+        _enable(None, exaile, None)
+
+def _enable(eventname, exaile, nothing):
+    '''
+        Called when plugin is loaded.
+    '''
+    pass
+
+def disable(exaile):
+    '''
+        Called when plugin is unloaded.  Remove Menu Item. Clean up.
+    '''
+    global SHUFFLE
+    if SHUFFLE:
+        SHUFFLE.stop()
+        SHUFFLE = None
+
+class Shuffle(xl.player.adapters.PlaybackAdapter):
+    def __init__(self, exaile):
+        self.exaile = exaile
+        self.do_shuffle = False
+        self.last_artists = []
+        self.tracks = list()
+        self.ban_repeat = 100 # ban an artist for x tracks
+        self.build_menu()
+
+    def build_menu(self):        
+        menu = xlgui.widgets.menu.check_menu_item('shuffle', ['plugin-sep'], 'Shuffle',
+            lambda *x: self.do_shuffle, lambda w, n, p, c: self.on_toggled(w))
+        xl.providers.register('menubar-tools-menu', menu)
+        xl.event.add_callback(self.on_playback_track_start, "playback_track_start", xl.player.PLAYER)
```

Comment:
This isn't part of building a menu, and you never remove the callback when the plugin is deactivated.